README


This archive contains the distribution Markdown-phpBB,
version 0.01:

  Turn markdown into phpBB code

This software is copyright (c) 2014 by Paul Fenwick.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Markdown::phpBB

0.01      2014-01-16 10:31:37 Australia/Melbourne

    * First release.


lib/Markdown/phpBB.pm
package Markdown::phpBB;

use 5.010;
use strict;
use warnings;
use autodie;
use Moose;

# ABSTRACT: Turn markdown into phpBB code
our $VERSION = '0.01'; # VERSION

use Markdent::Parser;
use Markdown::phpBB::Handler;


my $handler = Markdown::phpBB::Handler->new;

my $parser = Markdent::Parser->new(
    handler => $handler,
    dialect => 'GitHub',
);


sub convert {
    my ($self, $text) = @_;

    $parser->parse(markdown => $text);

    return $handler->result;
}

1;

__END__

=pod

=head1 NAME

Markdown::phpBB - Turn markdown into phpBB code

=head1 VERSION

version 0.01

=head1 SYNOPSIS

    my $md2php = Markdown::phpBB->new;

    my $phpbb = $md2php->convert($markdown);

=head1 DESCRIPTION

This converts (github-flavoured) markdown into phpBB / BBcode.

It uses L<Markdown::phpBB::Handler> and L<Markdent> to do the
heavy lifting.

=head1 METHODS

=head2 convert

    my $phpbb = $md2php->convert($markdown);

Takes a single string in markdown format, and returns the equivalent
string in phpBB / BBcode.

=head1 BUGS

Plenty. Report them or fix them at
L<http://github.com/pjf/Markdown-phpBB/issues>.

=head1 AUTHOR

Paul Fenwick <pjf@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Paul Fenwick.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Markdown/phpBB/Handler.pm
package Markdown::phpBB::Handler;

# ABSTRACT: Turn Markdown into phpBB code

use 5.010;
use strict;
use warnings;

use Moose;
use Data::Dumper;

our $VERSION = '0.01'; # VERSION


with 'Markdent::Role::Handler';

has _cached    => (is => 'rw', isa => 'Str', default => '');

our $DEBUG = $ENV{MD2PHPBBDEBUG} || 0;


sub handle_event {
    my ($self, $event) = @_;

    $self->_add( $self->_text_for_event($event) || "");
    
    return;
}

sub _add {
    my ($self, $text) = @_;

    warn "$text\n" if $DEBUG;

    $self->_cached( $self->_cached . $text );
}


sub result {
    my ($self) = @_;

    my $cached = $self->_cached();

    $self->_cached(""); # Clear cache;

    # Fix tags on wrong lines
    $cached =~ s{\n\[/li\]}{[/li]\n}g;
    $cached =~ s{\n\[/size\]\[/b\]}{[/size][/b]\n}g;

    $cached =~ s{\s*$}{\n};                   # Remove trailing whitespace

    return $cached;

}

my %tag = (
    document       => [    "",                           ""             ],
    paragraph      => [    "",                           "\n"           ],
    emphasis       => [ qw([i]                           [/i])          ],
    strong         => [ qw([b]                           [/b])          ],
    unordered_list => [   "[list]\n",                   "[/list]\n"     ],
    ordered_list   => [   "[list type=decimal]\n",      "[/list]\n"     ],
    list_item      => [ qw([li]                          [/li] )        ],
    link           => [ qw([url]                         [/url])        ],
    header         => [   "[b][size]",                  "[/size][/b]\n" ],
    code           => [   "[font=courier]",             "[/font]"       ],
);

my @heading_size = (36, 24, 18, 14, 12);

sub _text_for_event {
    my ($self, $event) = @_;

    my $name  = $event->event_name;
    my $start = $event->is_start;

    if ($name eq 'text')            { return $event->text; }
    if ($name eq 'horizontal_rule') { return "[hr]\n\n"    }

    if ($name eq 'code_block')      { return "[code]".$event->code."[/code]" }
    if ($name eq 'preformatted')    { return "[code]".$event->text."[/code]" }

    if ($name eq 'start_link' ) {
        my $url = $event->uri;
        return "[url=$url]";
    }

    if ($name eq 'image') {
        my $url = $event->uri;
        return "[spoiler][img]$url\[/img][/spoiler]"
    }


    if ($name eq 'start_header') {
        my $level = $event->level;
        my $size  = $heading_size[$level-1];
        return "[b][size=${size}pt]"
    }

    $name =~ s/^(?:start|end)_//;
    if ($tag{$name}) { return $tag{$name}[$start ? 0 : 1]; }

    # Oh noes! Something went wrong.

    use Data::Dumper;
    warn "Unknown markdown event: ". $event->event_name . "\n\n" . Dumper { $event->kv_pairs_for_attributes };

    return;

}

1;

__END__

=pod

=head1 NAME

Markdown::phpBB::Handler - Turn Markdown into phpBB code

=head1 VERSION

version 0.01

=head1 SYNOPSIS

    use Markdent::Parser;
    use Markdown::phpBB::Handler;

    my $handler = Markdown::phpBB::Handler->new;

    my $parser = Markdent::Parser->new(
        handler => $handler,
        dialect => 'GitHub',  # optional
    );

    $parser->parse(markdown => $md);

    my $phpBB = $handler->reseult;

=head1 METHODS

=head2 handle_event

Called by L<Markdent::Parser>. Takes an event and processes it.

=head2 result

    my $phpbb = $handler->result;

Returns the final string in phpBB after conversion.

Note that in the current version, calling this also resets the handler
state, so subsequent calls with return an empty string. Patches welcome
to fix this.

=head1 DESSCRIPTION

This is a L<Markdent::Role::Handler> which produces phpBB / BBcode
from Markdown.

It will emit a warning (but will continue) if it encounters events
it does not understand. Patches are very welcome.

=head1 SEE ALSO

L<Markdown::phpBB>, L<md2phpbb>, L<phpbb2md>, L<Markdent>

=head1 BUGS

Plenty. In particular, calling C<result> a second time will return an
empty string.

Report them or fix them at
L<http://github.com/pjf/Markdown-phpBB/issues>.

=head1 AUTHOR

Paul Fenwick <pjf@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Paul Fenwick.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Turn markdown into phpBB code",
  "AUTHOR" => "Paul Fenwick <pjf\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Markdown-phpBB",
  "EXE_FILES" => [
    "bin/md2phpbb",
    "bin/phpbb2md"
  ],
  "LICENSE" => "perl",
  "NAME" => "Markdown::phpBB",
  "PREREQ_PM" => {
    "Data::Dumper" => 0,
    "Markdent::Parser" => 0,
    "Markdent::Role::Handler" => 0,
    "Moose" => 0,
    "autodie" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {},
  "VERSION" => "0.01",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



